clippy_lints/src/too_owned.rs

```diff
@@ -0,0 +1,84 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_opt;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects needlessly owned `Cow` types.
+    ///
+    /// ### Why is this bad?
+    /// The borrowed types are usually more flexible (e.g. `&str` vs. `&String`).
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let wrogn: std::borrow::Cow<'_, String>;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let right: std::borrow::Cow<'_, str>;
+    /// ```
+    #[clippy::version = "1.85.0"]
+    pub OWNED_COW,
+    style,
+    "needlessly owned Cow type"
+}
+
+declare_lint_pass!(TooOwned => [OWNED_COW]);
+
+impl LateLintPass<'_> for TooOwned {
+    fn check_ty(&mut self, cx: &LateContext<'_>, ty: &hir::Ty<'_>) {
+        if let hir::TyKind::Path(qpath) = ty.kind
+            && let Some(cow_def_id) = cx.qpath_res(&qpath, ty.hir_id).opt_def_id()
+            && cx.tcx.is_diagnostic_item(sym::Cow, cow_def_id)
+            && let hir::QPath::Resolved(_, path) = qpath
+            && let [.., last_seg] = path.segments
+            && let Some(args) = last_seg.args
+            && let [_lt, carg] = args.args
+            && let hir::GenericArg::Type(cty) = carg
+            && let Some((span, repl)) = replacement(cx, cty)
+        {
+            span_lint_and_sugg(
+                cx,
+                OWNED_COW,
+                span,
+                "needlessly owned Cow type",
+                "use",
+                repl,
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
+fn replacement(cx: &LateContext<'_>, cty: &hir::Ty<'_>) -> Option<(Span, String)> {
+    if clippy_utils::is_path_lang_item(cx, cty, hir::LangItem::String) {
+        return Some((cty.span, "str".into()));
+    }
+    if clippy_utils::is_path_diagnostic_item(cx, cty, sym::Vec) {
+        return if let hir::TyKind::Path(hir::QPath::Resolved(_, path)) = cty.kind
+            && let [.., last_seg] = path.segments
+            && let Some(args) = last_seg.args
+            && let [t, ..] = args.args
+            && let Some(snip) = snippet_opt(cx, t.span())
+        {
+            Some((cty.span, format!("[{snip}]")))
+        } else {
+            None
+        };
+    }
+    for (diag, repl) in [
+        (sym::cstring_type, "std::ffi::CStr"),
```

Comment:
This one should probably use `clippy_utils::std_or_core()` as its prefix, or the suggestion will be incorrect in a `no_std` context (with the `alloc` crate for `alloc::ffi::CString`).

Reply:
I added some code to distinguish no_std code. Good catch!